Add back DistributedDataParallel types that were lost when pyi was removed

mauvilsa opened the pull request:
When the stub file `nn/parallel/distributed.pyi` was removed (#88701), some types that existed are no longer available. This pull request adds them back.

Just for reference, these types are used in pytorch-lightning's LightningCLI. Command line interfaces are created automatically, and having type hints make them nicer.



cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

mauvilsa committed:
Add again the DistributedDataParallel types that were lost when distributed.pyi was removed.

mauvilsa commented:
@pytorchbot label "topic: not user facing"

kwen2501 reviewed (dismissed):
Looks okay to me

mauvilsa committed:
Address lintrunner errors

mauvilsa commented:
[lintrunner-noclang](https://github.com/pytorch/pytorch/actions/runs/11064358980/job/30746336842#logs) gave mypy errors, so I pushed a new commit to fix them.

mauvilsa commented on torch/nn/parallel/distributed.py:
```diff
@@ -625,24 +637,26 @@ class DistributedDataParallel(Module, Joinable):
+        output_device: Optional[Union[int, device]] = None,
+        dim: int = 0,
+        broadcast_buffers: bool = True,
+        process_group: "Optional[ProcessGroup]" = None,
```
I changed this type, so now it is different than in the `pyi` file. From the code it seems it should be like this. Though I had to add a `# type: ignore[assignment]` to line 679.

mauvilsa commented on torch/nn/parallel/distributed.py:
```diff
@@ -625,24 +637,26 @@ class DistributedDataParallel(Module, Joinable):
         delay_all_reduce_named_params=None,
         param_to_hook_all_reduce=None,
         mixed_precision: Optional[_MixedPrecision] = None,
         device_mesh=None,
```
`device_mesh` didn't exist when `distributed.pyi` was removed, which is why I didn't add it. But I could add a type for it if you like.

mauvilsa replied:
Added it.

mauvilsa committed:
Add type for device_mesh

kwen2501 commented:
@pytorchbot merge

mauvilsa committed:
Make reducer type hint a string

mauvilsa commented:
I noticed that tests on mac arm64 were failing because of the reducer type added. Made the type hint explicitly a string to avoid these failures.

cyyever commented:
@pytorchbot merge

cyyever commented:
@pytorchbot rebase

cyyever commented:
@pytorchbot merge

PaliC commented:
@pytorchbot revet -m "this pr is causing typecheck errors internally" -c "ghfirst"
@kwen2501 can you look into fixing the type check errors so we can merge this

mauvilsa committed:
Change device_ids type from Sequence to List

mauvilsa commented:
> Stack Trace for your reference: Incompatible parameter type [6]: In call `nn.parallel.distributed.DistributedDataParallel.__init__`, for argument `device_ids`, expected `Optional[Sequence[TypeUnion[int, device]]]` but got `Optional[List[Optional[device]]]`.

I have changed the type for `device_ids` from `Sequence` to `List` just based on the error message. Though, would be nice to know how to reproduce that error. `lintrunner` was not giving such an error before, so I don't really know if a similar thing would happen again. @jovianjaison any advice?

mauvilsa commented:
> I have changed the type for `device_ids` from `Sequence` to `List` just based on the error message. Though, would be nice to know how to reproduce that error locally. `lintrunner` was not giving such an error before, so I don't really know if a similar thing would happen again. @jovianjaison, @PaliC any advice?

@kwen2501 maybe you can let me know where it failed and how to reproduce? Anyway, I changed the code and you could look again.

kwen2501 commented:
Thanks for the change. Let's just try merge again.

kwen2501 commented:
@pytorchbot rebase

kwen2501 commented:
@pytorchbot merge

kwen2501 reviewed (dismissed):
LGTM

kwen2501 commented:
@pytorchbot merge

cyyever commented:
@pytorchbot merge

izaitsevfb commented:
@pytorchbot revert -m "breaking typechecks in meta code" -c ghfirst

I'm not sure that 
```
       device_ids: Optional[List[Union[int, device]]] = None,
        output_device: Optional[Union[int, device]] = None,
```
are correct types to have here.

`List` type is invariant in python, which means that  neither  `List[int]` nor `List[device]` are subtypes of `Optional[List[Union[int, device]]]`.

I'd suggest to change the type signature to something like:
`Optional[Union[List[int], List[device]]]` or `Optional[Union[List[int], List[device], List[Union[int, device]]]]` 
to support existing code that passes e.g. list of device indices.


actual errors that we're getting:
```
Found type errors!
fbcode/aps_models/ads/common/ads_data_parallel_wrapper.py:137:16 Incompatible parameter type [6]: In call `torch.nn.parallel.distributed.DistributedDataParallel.__init__`, for argument `device_ids`, expected `Optional[List[Union[int, device]]]` but got `Optional[List[device]]`.
```

mauvilsa committed:
Fix device_ids type

mauvilsa committed:
Add type to init_sync and address linter errors

mauvilsa commented:
I had forgotten about this pull request, sorry. @izaitsevfb thank you for the info. I have updated the type. @kwen2501 shall we try this again.

mauvilsa commented:
@pytorchbot remove label Stale

mauvilsa commented:
@pytorchbot label --help

mauvilsa commented:
@pytorchbot --help

cyyever commented:
@pytorchbot rebase

cyyever commented:
@mauvilsa @kwen2501 Do you need to import it to check whether it works in META?

mauvilsa commented:
@cyyever this has been merged and reverted twice because tests in meta fail. With the rebase from today now there are new lintrunner errors and warnings. I will fix them.

mauvilsa committed:
Fix new lintrunner errors and warnings

mauvilsa committed:
Revert new union syntax

mauvilsa committed:
Merge branch 'main' into types-DistributedDataParallel
